internal/compiler-bridge/src/main/scala/xsbt/ExtractUsedNames.scala

```diff
@@ -49,29 +52,39 @@ import Compat._
 class ExtractUsedNames[GlobalType <: CallbackGlobal](val global: GlobalType) extends Compat with ClassName with GlobalHelpers {
   import global._
 
-  def extract(unit: CompilationUnit): Iterable[(String, Iterable[String])] = {
+  def extract(unit: CompilationUnit): HashMap[String, ArrayList[String]] = {
     val tree = unit.body
     val traverser = new ExtractUsedNamesTraverser
     traverser.traverse(tree)
     val namesUsedAtTopLevel = traverser.namesUsedAtTopLevel
 
-    if (namesUsedAtTopLevel.nonEmpty) {
+    if (!namesUsedAtTopLevel.isEmpty) {
       val responsible = firstClassOrModuleDef(tree)
       responsible match {
         case Some(classOrModuleDef) =>
           val sym = classOrModuleDef.symbol
           val firstClassSymbol = if (sym.isModule) sym.moduleClass else sym
           val firstClassName = className(firstClassSymbol)
-          traverser.usedNamesFromClass(firstClassName) ++= namesUsedAtTopLevel
+          traverser.usedNamesFromClass(firstClassName).addAll(namesUsedAtTopLevel)
         case None =>
           reporter.warning(unit.position(0), Feedback.OrphanNames)
       }
     }
 
-    traverser.usedNamesFromClasses.map { tpl =>
-      // NOTE: We don't decode the full class name, only dependent names.
-      tpl._1.toString.trim -> tpl._2.map(_.decode.trim)
+    val result = new HashMap[String, ArrayList[String]]()
+
+    val it = traverser.usedNamesFromClasses.entrySet().iterator()
+    while (it.hasNext) {
+      val usedName = it.next()
+      val usedNameKey = usedName.getKey.toString.trim
+      val usedNameValues = usedName.getValue.iterator()
+      val uses = new ArrayList[String]()
```

Comment:
Interestingly, it seems that we're intiializing an array list here every time we iterate through the class name entries. It would be nice if we could remove it.
  
It happens that `HashSet` has a constant `size` method, so do you think we could use that size to create an array of that concrete length and then change the return type of `extract`? This would buy us the fact that we are not paying the cost of expanding the array list when the capacity is full because we know that size a priori. It can also help us with some spacial cache locality.